### PR TITLE
Add more elements in the sidebar

### LINKED ISSUE #45740: Traits should list their methods in the sidebar as well (and not their "traits")
Take a look at the current https://doc.rust-lang.org/nightly/std/clone/trait.Clone.html for a better understanding. :)

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #45740.

r? @rust-lang/docs

### CODE REVIEW

File: src/test/rustdoc/sidebar-items.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#![crate_name = "foo"]
+
+// @has foo/trait.Foo.html
+// @has - '//*[@class="sidebar-title"]/a' 'Required Methods'
+// @has - '//*[@class="sidebar-links"]/a' 'bar'
+// @has - '//*[@class="sidebar-title"]/a' 'Provided Methods'
+// @has - '//*[@class="sidebar-links"]/a' 'foo'
+// @has - '//*[@class="sidebar-title"]/a' 'Associated Constants'
+// @has - '//*[@class="sidebar-links"]/a' 'BAR'
+// @has - '//*[@class="sidebar-title"]/a' 'Associated Types'
```
Comment (kennytm) [inline src/test/rustdoc/sidebar-items.rs]:
```
[00:59:22] 14: @has check failed
[00:59:22] 	`XPATH PATTERN` did not match
[00:59:22] 	// @has - '//*[@class="sidebar-title"]/a' 'Required Methods'
[00:59:22] 16: @has check failed
[00:59:22] 	`XPATH PATTERN` did not match
[00:59:22] 	// @has - '//*[@class="sidebar-title"]/a' 'Provided Methods'
[00:59:22] 18: @has check failed
[00:59:22] 	`XPATH PATTERN` did not match
[00:59:22] 	// @has - '//*[@class="sidebar-title"]/a' 'Associated Constants'
[00:59:22] 20: @has check failed
[00:59:22] 	`XPATH PATTERN` did not match
[00:59:22] 	// @has - '//*[@class="sidebar-title"]/a' 'Associated Types'
[00:59:22] 31: @has check failed
[00:59:22] 	`XPATH PATTERN` did not match
[00:59:22] 	// @has - '//*[@class="sidebar-title"]/a' 'Fields'
[00:59:22] 41: @has check failed
[00:59:22] 	`XPATH PATTERN` did not match
[00:59:22] 	// @has - '//*[@class="sidebar-title"]/a' 'Variants'
[00:59:22] 50: @has check failed
[00:59:22] 	`XPATH PATTERN` did not match
[00:59:22] 	// @has - '//*[@class="sidebar-title"]/a' 'Fields'
[00:59:22] 52: @has check failed
[00:59:22] 	`XPATH PATTERN` did not match
[00:59:22] 	// @has - '//*[@class="sidebar-links"]/a' 'u'
```

**Comment by GuillaumeGomez** [comment]
Working!

File: src/librustdoc/html/render.rs [hunk 2]
```diff
@@ -3547,6 +3547,10 @@ fn small_url_encode(s: &str) -> String {
      .replace("]", "%5D")
 }
 
+fn small_html_encode(s: &str) -> String {
```
Comment (ollie27) [inline src/librustdoc/html/render.rs]: Why aren't you using `Escape`? `&`s appear in type parameters for impls so need to be escaped which `Escape` will handle but this doens't.

**Review by ollie27 — CHANGES_REQUESTED** [review]

File: src/librustdoc/html/render.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render.rs]: I didn't understand the second part of your comment (and to answer the first, I didn't know it existed, I'll update!).

Comment (QuietMisdreavus) [reply src/librustdoc/html/render.rs]: [`impl<'a> Add<&'a i32> for i32`](https://doc.rust-lang.org/nightly/std/primitive.i32.html#impl-Add%3C%26%27a%20i32%3E) is a valid thing to attempt to print, so the `&` in there should be escaped properly in the html.

**Comment by QuietMisdreavus** [comment]
r=me with @ollie27's comment (and with travis)

**Comment by GuillaumeGomez** [comment]
Updated, let's wait for travis now.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by kennytm** [comment]
@bors r=QuietMisdreavus

Travis is green and the code is now using `Escape`.

**Pull request merged**